### DISCUSSION TITLE
about Changing the size of a VARCHAR field

### DISCUSSION THREAD

**Discussion by ian-lss** [milvus-io/milvus#34478] [Q&A and General discussion] [answered]
Hello

I’m not sure if I missed it in the documentation, 
but it seems that changing the size of a VARCHAR column is not supported in pymilvus. 
If that’s the case, how can I change the field size while the service is online?

**Comment by yhmo** [comment]
max_length of a varchar field cannot be changed once the collection is created.
The max value of max_length is 65535. If you are unsure which value is proper, you can set max_length=65535 when creating the collection.

**Comment by ian-lss** [comment]
@yhmo 

Thank you for your response.
If I always set the length of all VARCHAR fields to 65535, 
what potential issues could arise?

**Reply by yhmo** [reply]
No potential issue. The occupied space(memory/disk) of a varchar string depends on its real length.

**Reply by ian-lss** [reply]
I have one last question.
If there are no potential issues, why does Milvus require specifying the size for VARCHAR type instead of defaulting it to 65535? 
Is it just to align with the interface of other databases?

**Reply by yhmo** [reply]
It is just like the VARCHAR of MySQL, variable string requires a max_length value to limit the length.
There is another type TEXT in MySQL, long length string without any limitation. Milvus doesn't support TEXT.

**Reply by ian-lss** [reply]
The reason for keeping the VARCHAR size tight in MySQL is that the buffer pool scans the storage space based on the VARCHAR size, not the actual data length. 
Of course, the actual storage uses only the actual data length. 
Additionally, since the VARCHAR length can be changed according to the service situation, it is possible to do so.

If Milvus scans the storage space for the query node based on the actual data length, then guiding developers to use 65535 for the max_length should not be an issue. 
Since the length can be changed at any time from a service perspective, I plan to set it to 65535 as long as there are no performance issues. This is why I asked the question.

**Reply by yhmo** [reply, marked as answer]
Actually, the max_length value affects the internal logic of milvus (such as determining the size of insert buffer).

In ths source code of pkg/util/typeutil/schema.go
- EstimateSizePerRecord is used to determine the size of the insert buffer
- EstimateMaxSizePerRecord is used to determine the size of the read buffer of bulkinsert

max_length is calculated by these methods. So, different values will affect the internal behavior of milvus.  From the source code, I didn't see it affects search/query performance.
An accurate max_length is welcome, 65535 is also ok if users don't like the limitation.

https://github.com/milvus-io/milvus/blob/9b37d3f5178830341bd15974dd09c4a049bcadf4/pkg/util/typeutil/schema.go#L105
https://github.com/milvus-io/milvus/blob/9b37d3f5178830341bd15974dd09c4a049bcadf4/pkg/util/typeutil/schema.go#L50

**Reply by ian-lss** [reply]
@yhmo 
Thank you for your response.
I fully understand your detailed explanation.

**Comment by xiaofan-luan** [comment]
/assign @PowderLi 
please help on the feature of changing varchar length
